### ISSUE TITLE
Selective backpropagation over a graph

### ISSUE THREAD

**Issue by surojitiitg** [pyg-team/pytorch_geometric#8384]
### 🚀 The feature, motivation and pitch

We are working on a research project, where we need to prevent back propagation between some nodes while allowing for forward propagation. This represents a graph that is dynamic in nature and allows for eliminating links/connections in the backward propagation, depending on the scenario. I have attached an image in the "Additional Context" section to explain the scenario further.

### Alternatives

While posting this problem on slack channel, we discussed about using detach (). However, I was not able to follow along how to use it given we have to use batching while processing the data and cannot simply detach a batch tensor. Furthermore,  we want to have the ability to detach the backward pass between node B and C but keep the backward pass between Node C and D (This is just a sample example).

### Additional context

I am attaching here a diagram to represent what I am referring to.

<img width="1014" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/6539795/684af115-aae3-448e-bec5-0ce2c7241e9e">

**Comment by rusty1s** [comment]
Thanks for the detailed explanation. In general, this is hard to achieve. One way to resolve it would be to also maintain a `edge_mask` about the edges you want to maintain for gradient computation. Then, inside `message`, you could do
```python
def message(self, x_j, edge_mask):
    masked_x_j = x_j.clone()
    masked_x_j[~edge_mask] = 0
    return x_j.detach() - masked_x_j.detach() + masked_x_j
```
similar to what is done in gumbel softmax (see [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.gumbel_softmax.html))

**Comment by surojitiitg** [comment]
Thanks for the explanation. Based on the code snippet is the following true: 
(1)"Message" function has x_j, which is a vector having size equal to number of links. edge_mask has a size equal to number of links with 1's at the edges for which the back propagation is desired and zero other places. 
(2)In the second line of the function, you zero out the messages corresponding to the edge inks for which backpropagation tracking is not desired
(3)I am not able to interpret the return statement very well and getting lost here. Basically, a tensor is returned that is equal to x_j for which we still have the gradient tracking for all messages (Adding an non detached tensor to a detached tensor makes the sum non detached). Is my understanding of the return statement, right? or am I missing your explanation completely.

I think, conceptually I am not able to understand how to detach only certain portions of the x_j vector. If you can please provide some clarifications on these points, that will help a lot.

**Comment by rusty1s** [comment]
I didn't have tested (3), but it should work.

The idea is to return `x_j` in the forward, but return `masked_x_j.grad` in the backward. You are right that the sum is still non-detached (which is desired), but during backpropagation, grads are only backpropagated to `masked_x_j` (since `x_j` is detached). `masked_x_j.grad` will be then propagated to `x_j` but only to valid entries according to `edge_mask`.
